### PR TITLE
DDL: createtable not support empty name index 

### PR DESCRIPTION
**Pull request by watchpoints**
### What problem does this PR solve?

Issue Number: #18149 
create table  not support  create empty name index

after parser PR: pingcap/parser/pull/1059
then update go.mod.

Problem Summary:

### What is changed and how it works?

change token IndexName from ident to item type
use a NullIndexName struct to represent a empty index to distinguish from a anonymous index

### Related changes
- pingcap/parser

### Check List 

Tests 

- Unit test
go test -check.f TestValidator 
go test -check.f TestCreateTable

- Manual test (add detailed scripts or steps below)
In TiDB
~~~
mysql> CREATE TABLE t(x INT, KEY ``(x));
ERROR 1280 (42000): Incorrect index name '
~~~
In MySQL:
~~~
mysql> CREATE TABLE t1(x INT, KEY ``(x));
ERROR 1280 (42000): Incorrect index name ''
~~~
It has the same action with MySQL

Side effects

### Release note 

Fix not support empty index when createtableding tables ,but  anonymous index  also support

### CODE REVIEW

File: ddl/db_test.go [hunk 1]
```diff
@@ -2621,6 +2621,14 @@ func (s *testSerialDBSuite) TestCreateTable(c *C) {
 	tk.MustGetErrCode(failSQL, errno.ErrTableOptionInsertMethodUnsupported)
 	tk.MustExec("drop table x;")
 	tk.MustExec("drop table y;")
+
+	// For issue #18149
+	tk.MustExec("use test")
+	failSQL = "CREATE TABLE t(a INT, KEY ``(a));"
+	tk.MustGetErrCode(failSQL, errno.ErrWrongNameForIndex)
+	_, err = tk.Exec("CREATE TABLE t(a INT, KEY ``(a));")
```
Comment (xhebox) [inline ddl/db_test.go:2629]: I don't think you need to check error again if you have checked it by `MustGetErrCode`.

Comment (xhebox) [inline ddl/db_test.go:2629]: I don't think you need to check error again if you have checked it by MustGetErrCode.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -1182,6 +1187,10 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 
 	// Set empty constraint names.
 	for _, constr := range constraints {
+		err := checkEmptyIndexInfo(constr.Name, constr.IsEmptyIndex)
```
Comment (xhebox) [inline ddl/ddl_api.go:1190]: Is there a need to check it again? I think statements here are all preprocessed in planner.

**Review by xhebox — COMMENTED** [review]
You can update go.mod first to run all tests:

```
@ti-srebot /run-all-tests
```

**EDIT**: From the parser PR, `IsEmptyIndex` is true only when `len(str) == 0`, which means `name == ""` holds if `IsEmptyIndex == true`. The check for `name == ""` seems redundant.

**Comment by watchpoints** [comment]
I delete the tidb ,and  fork  again,  I can not  commit this pr ,so  it is closed.

**Pull request closed without merging**